src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -902,6 +902,25 @@ public SolutionState WithProjectReferences(ProjectId projectId, IEnumerable<Proj
             return this.ForkProject(newProject, newDependencyGraph: _dependencyGraph.WithProjectReferences(projectId, projectReferences.Select(p => p.ProjectId)));
         }
 
+        /// <summary>
+        /// Creates a new solution instance with the project documents in the order by the specified document ids.
+        /// The specified document ids must exist within the specified project and have the same count as the number of documents in the specified project.
```

Comment:
Ditto with earlier comment: the count check is just a quick way to catch obvious screwups; the real contract here is you can't remove/add.